### PR TITLE
Warn user of all tabs with unsaved progress when window is closed.

### LINKED ISSUE #273: save before quit warning for unsaved files
it would be nice to be warned about unsaved files when app quit.

### PR DESCRIPTION
**Pull request by BugsNash**
This patch iterates through each open tab and prompts the user with the save file dialog for each tab with unsaved progress before closing the window. This fixes #273.

This has already been reviewed in a separate PR to my own repo at https://github.com/BugsNash/text-app/pull/1

### CODE REVIEW

File: js/tabs.js [hunk 1]
```diff
@@ -326,6 +309,54 @@ Tabs.prototype.openFiles = function() {
...
+/**
+ * @param {function()} callback
+ * Invoke the save dialog for all tabs with unsaved progress. Does not close any tabs.
+ */
+Tabs.prototype.promptAllUnsaved = function(callback) {
+  this.promptAllUnsavedFromIndex_(0, callback);
+}
+
+Tabs.prototype.promptAllUnsavedFromIndex_ = function(i, callback) {
+  if (this.tabs_.length > i) {
```
Comment (beaufortfrancois) [inline js/tabs.js]: Nit: How about bailing out early?

```js
if (this.tabs_.length === 0) {
  callback();
  return;
}
...
```

Comment (BugsNash) [reply js/tabs.js]: doesn't the current code already do that? if tabs_.length is 0, we immediately hit the else block below that calls callback()

File: js/tabs.js [hunk 2]
```diff
@@ -322,6 +309,54 @@ Tabs.prototype.openFiles = function() {
...
+  }
+}
+
+Tabs.prototype.promptSave_ = function(tab, callbackShowDialog) {
+  var filename = tab.getName();
+  if (filename.length > 25) {
+    filename = filename.slice(0, 22).concat('...');
+  }
+
+  this.dialogController_.setText(
```
Comment (E-T-) [inline js/tabs.js]: A side note --- this feels like something that a Builder pattern or some restructuring could help with. Not something for this patch, just an observation that this is a little difficult to read atm.

File: js/tabs.js [hunk 3]
```diff
@@ -322,6 +309,54 @@ Tabs.prototype.openFiles = function() {
...
+  }
+  else {
+    callback();
+  }
+}
+
+Tabs.prototype.promptSave_ = function(tab, callbackShowDialog) {
+  var filename = tab.getName();
+  if (filename.length > 25) {
+    filename = filename.slice(0, 22).concat('...');
```
Comment (E-T-) [inline js/tabs.js]: Can we use ellipses here?

File: js/tabs.js [hunk 4]
```diff
@@ -322,6 +309,54 @@ Tabs.prototype.openFiles = function() {
...
+        if (answer === 'yes') {
+          this.save(tab, false /* close */);
+        } else if (answer === 'cancel') {
+          return;
+        }
+        this.promptAllUnsavedFromIndex_(i + 1, callback);
+      }.bind(this));
+    }
+  }
+  else {
```
Comment (E-T-) [inline js/tabs.js]: I would suggest...

if (i >= this.tabs_.length) {
  callback();
  return;
}

Which will unindent the other block and make it simpler to read.

File: js/tabs.js [hunk 5]
```diff
@@ -322,6 +309,54 @@ Tabs.prototype.openFiles = function() {
       this.openFileEntry.bind(this));
 };
 
+/**
+ * @param {function()} callback
+ * Invoke the save dialog for all tabs with unsaved progress. Does not close any tabs.
+ */
+Tabs.prototype.promptAllUnsaved = function(callback) {
+  this.promptAllUnsavedFromIndex_(0, callback);
```
Comment (E-T-) [inline js/tabs.js]: FMI, why not a for loop here, and then you could combine promptAllUnsavedFromIndex_ and promptAllUnsaved_ into one function?

File: js/tabs.js (hunk 3 above)
Comment (BugsNash) [reply js/tabs.js]: AAAArrrrhhhhhblj rebasing error again *head smack* thanks for catching this

File: js/tabs.js (hunk 2 above)
Comment (BugsNash) [reply js/tabs.js]: Ack

File: js/tabs.js (hunk 4 above)
Comment (BugsNash) [reply js/tabs.js]: Done.

File: js/tabs.js (hunk 5 above)
Comment (BugsNash) [reply js/tabs.js]: I thought promptAllUnsaved was nicer as a public function, happy to change if you disagree :)

**Comment by BugsNash** [comment]
Reopened as https://github.com/GoogleChromeLabs/text-app/pull/336 because rebasing and merging seems to have messed things up and my pushed changes weren't being reflected in this PR

**Pull request closed without merging**